### PR TITLE
fix wrong reprojction to wrf domain

### PR DESCRIPTION
**Pull request by matthiasdemuzere**
Addressing Issue #56 

**Issue**: Reprojection from the high-res LCZ grid to the lower-res WRF grid resulted in errors for larger domains.
**Reason**: WRF has an irregular latlong grid, and this was not taken into account when reprojecting.
**Solution**: Inspired by the blog information provided by Fabien Maussion [here](https://fabienmaussion.info/2018/01/06/wrf-projection/), the procedure is not completely changed:

- the `create_wrf_gridinfo()` routine that wrote a temporary .tif file, containing WRF's grid info, is now removed.
- instead a new routine `_get_wrf_grid_info()` is introduced to get a proper description of the WRF grid, in its native Lambert Conformal Conic projection.
- this routine provide the proper `crs` and `transform` values needed to Reproject the lcz info to the wrf grid.

### CODE REVIEW

**Comment by matthiasdemuzere** [comment]
Ok, as far as I can see this all works now:
- new reprojection procedure seems to work well.
- tests have been updated and pass
- new test has been added to check the `_get_wrf_grid_info()` routine
- relevant sample_data/ and testing/ files have been updated

@dargueso @theendlessriver13: perhaps you can review? 

If accepted, I guess we should bump the version number, and update pypi. 
Especially for the larger WRF domains this change will have some impact ....

**Comment by dargueso** [comment]
@matthiasdemuzere, sure, I can check it out, but won't be able to do so before Monday.
Hope that works.

**Comment by matthiasdemuzere** [comment]
I realized there is another level of complication that we did not yet take care of.

Geo_em files can be on different projections. This is information is stored in MAP_PROJ.

for sample data:
```
>>> ds = xr.open_dataset(geo_em.d04.nc)
>>> print(f'{case}: MAP_PROJ = {ds.MAP_PROJ}')
sample_data: MAP_PROJ = 6
```

for Italy data from pippo2013:
```
>>> ds = xr.open_dataset(geo_em.d04.nc)
>>> print(f'{case}: MAP_PROJ = {ds.MAP_PROJ}')
sample_data: MAP_PROJ = 1
```

`MAP_PROJ` can have the following values:
- 1: Lambert, 
- 2: polar stereographic, 
- 3: mercator, 
- 6: lat-lon 

More info:
- [https://www2.mmm.ucar.edu/wrf/users/namelist_best_prac_wps.html#map_proj](https://www2.mmm.ucar.edu/wrf/users/namelist_best_prac_wps.html#map_proj)
- [http://140.112.69.65/research/coawst/COAWST_TUTORIAL/training_2019/monday/werner_wps.pdf](http://140.112.69.65/research/coawst/COAWST_TUTORIAL/training_2019/monday/werner_wps.pdf)

So far this new code caters for 1-Lambert only.

So a check is needed within `_get_wrf_grid_info()`, to:
- get the MAP_PROJ value
- assign the corresponding wrf_proj string

@dargueso not sure if you feel like adding this?

**Comment by dargueso** [comment]
@matthiasdemuzere, sorry, I probably wasn't clear enough in my last messages. I thought you were aware of that. When I discussed the issue (https://github.com/matthiasdemuzere/w2w/issues/56#issuecomment-1139404336), I had this multiple projection option in mind. You have to add the rotated-pole option to the above. 

I think we can indeed do most of the work with salem or wrf-python. WRF-python reads all possible wrf projections and makes it easy to reproject after.

**Comment by matthiasdemuzere** [comment]
So far I developed a solution without additional packages. Could be good to keep it this way, as e.g. `salem` is somewhat maintained but not further developed ...

Let me add these other ones. 
For MAP_PROJ = 6, I'll have to check how this should be done.

**Comment by matthiasdemuzere** [comment]
Ok, I now added the different type of projections.

MAP_PROJ = 1 was tested with the Italy data. That looks fine, let's see if Pippo2013 can confirm this.

MAP_PROJ = 6 was tested with input from 
[costa_del_sol_input.zip](https://github.com/matthiasdemuzere/w2w/files/8832307/costa_del_sol_input.zip)

Unfortunately, the projection does not seem to be correct yet. So this needs more work.
@dargueso: perhaps you can have a look when you have some time? I started a review in the relevant section ...

image

File: w2w/w2w.py [hunk 1]
```diff
@@ -508,24 +500,87 @@ def wrf_remove_urban(
     dst_data.to_netcdf(info.dst_nu_file)
 
 
-# Make WRF grid info available for Resampler (tmp file)
-def create_wrf_gridinfo(info: Info) -> None:
+# Get WRF grid info for Resampler
+# Inspired by: https://fabienmaussion.info/2018/01/06/wrf-projection/
+def _get_wrf_grid_info(info: Info) -> Dict[str, Dict[str, Any]]:
```
Comment (theendlessriver13) [inline w2w/w2w.py:505]: This should be:
```suggestion
def _get_wrf_grid_info(info: Info) -> Dict[str, Any]:
```
That's how the dict looks like:
```python
{
    'crs': '+proj=eqc +lat_ts=0 +lat_0=0 +lon_0=5.84999990463257 +x_0=0 +y_0=0 '
        '+R=6370000 +units=m +no_defs',
    'transform': Affine(1111.7747802734375, 0.0, -838835.115342933,
       0.0, 1111.7747802734375, 4577176.609447704)
}
```

File: w2w/w2w.py [hunk 2]
```diff
@@ -508,24 +500,87 @@ def wrf_remove_urban(
...
+    dx, dy = dst_data.DX, dst_data.DY
+    nx, ny = dst_data.dims['west_east'], dst_data.dims['south_north']
+
+    # Down left corner of the domain
+    x0 = -(nx - 1) / 2.0 * dx + e
+    y0 = -(ny - 1) / 2.0 * dy + n
+
+    wrf_transform = Affine.translation(x0 - dx / 2, y0 - dy / 2) * Affine.scale(dx, dy)
+
+    wrf_grid_info: Dict[str, Dict[str, Any]] = {
```
Comment (theendlessriver13) [inline w2w/w2w.py:578]: We don't need a type annotation here
```suggestion
    wrf_grid_info = {
```

File: tests/w2w_test.py [hunk 3]
```diff
@@ -357,24 +355,28 @@ def test_wrf_remove_urban_output_already_exists_is_overwritten(tmpdir, info_mock
...
     )
-    create_wrf_gridinfo(info)
-    assert os.listdir(tmpdir) == ['dst_gridinfo.tif']
-    tif = rxr.open_rasterio(info.dst_gridinfo)
-    assert tif.rio.crs.to_proj4() == '+init=epsg:4326'
-    assert tif.rio.transform() == Affine(
-        0.01000213623046875,
+    wrf_grid_info = _get_wrf_grid_info(info)
+    assert type(wrf_grid_info) == dict
+    assert list(wrf_grid_info.keys()) == ['crs', 'transform']
```
Comment (theendlessriver13) [inline tests/w2w_test.py:366]: I don't think we need to test this, mypy does all the type checks with the function's type annotation. The keys are already tested with the asserts below, so we can also remove this
```suggestion
```

File: tests/w2w_test.py [hunk 4]
```diff
@@ -357,24 +355,28 @@ def test_wrf_remove_urban_output_already_exists_is_overwritten(tmpdir, info_mock
...
+    crs_string = (
+        '+proj=eqc +lat_ts=0 +lat_0=0 +lon_0=5.84999990463257 '
+        '+x_0=0 +y_0=0 +R=6370000 +units=m +no_defs'
+    )
+    assert wrf_grid_info['crs'] == crs_string
+    print(wrf_grid_info['transform'])
```
Comment (theendlessriver13) [inline tests/w2w_test.py:372]: no need to print
```suggestion
```

**Review by theendlessriver13 — COMMENTED** [review]
Looking very good, nice that this is done, thank you!
A few final questions:
- can you elaborate a little what has changed in the netcdf files - sample and testing data?
- Why are so many hard coded values in the tests changed, all due to the different projection and slightly moved pixels? What does this mean for data generated with the `0.2.0` version are they all slightly off?
- Would you classify this patch as a "bug fix" or a "new feature" so should this become version `0.2.1` or `0.3.0`?

**Comment by matthiasdemuzere** [comment]
> * can you elaborate a little what has changed in the netcdf files - sample and testing data?
The netcdf files in `sample_data/` are updated with the correct re-projected for the LCZ classes and urban canopy parameters, using this new implementation. As a result, the LCZ information in `LCZ_extent.nc` and `LCZ_params.nc` is slightly different.
The same is true for some of the files in `testing/

> * Why are so many hard coded values in the tests changed, all due to the different projection and slightly moved pixels? What does this mean for data generated with the `0.2.0` version are they all slightly off?
The previous reprojection was slightly off, as I was treating the WRF grid as a regular grid, but it is not.
The WRF grid can have 4 projection types (MAP_RPOJ 1, 2, 3 and 6) which are now all properly used when reprojecting the higher-res LCZ data onto the lower-res WRF grid.
That indeed means that the data generated with the `0.2.0` version is slightly off. Impact will be minimal for smaller domains, but more pronounced for bigger domains.

> * Would you classify this patch as a "bug fix" or a "new feature" so should this become version `0.2.1` or `0.3.0`?
This patch fixes a bug in the reprojection from high-res LCZ to low-res WRF.
Yet since outputs are changing between these two versions, it is probably better to assign this as a new feature, to bump the version from `0.2.1` to `0.3.0`?

**Review by theendlessriver13 — APPROVED** [review]
Nice, looking very good! Great that you were able to fix this. I'll make a new release to PyPi!

**Pull request merged**